iceberg/iceberg-handler/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java

```diff
@@ -0,0 +1,336 @@
+            .build();
+        break;
+      case ORC:
+        records =
+          ORC.read(inputFile)
+            .project(schema)
+            .createReaderFunc(fileSchema -> GenericOrcReader.buildReader(schema, fileSchema))
+            .build();
+        break;
+      case AVRO:
+        records = Avro.read(inputFile).project(schema).createReaderFunc(DataReader::create).build();
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported file format:" + fileFormat.name());
+    }
+
+    return CloseableIterable.transform(
+      records, PartitionStatsHandler::recordToPartitionStatsRecord);
+  }
+
+  private static FileFormat fileFormat(String fileLocation) {
+    return FileFormat.fromString(fileLocation.substring(fileLocation.lastIndexOf(".") + 1));
+  }
+
+  private static OutputFile newPartitionStatsFile(Table table, long snapshotId) {
+    FileFormat fileFormat =
+        fileFormat(
+        table.properties().getOrDefault(DEFAULT_FILE_FORMAT, DEFAULT_FILE_FORMAT_DEFAULT));
+    return table
+      .io()
+      .newOutputFile(
+        ((HasTableOperations) table)
+          .operations()
+          .metadataFileLocation(
+            fileFormat.addExtension(
+              String.format(Locale.ROOT, "partition-stats-%d", snapshotId))));
+  }
+
+  private static PartitionStatsRecord recordToPartitionStatsRecord(Record record) {
+    PartitionStats stats =
+        new PartitionStats(
+        record.get(Column.PARTITION.id(), StructLike.class),
+        record.get(Column.SPEC_ID.id(), Integer.class));
+    stats.set(Column.DATA_RECORD_COUNT.id(), record.get(Column.DATA_RECORD_COUNT.id(), Long.class));
+    stats.set(Column.DATA_FILE_COUNT.id(), record.get(Column.DATA_FILE_COUNT.id(), Integer.class));
+    stats.set(
+        Column.TOTAL_DATA_FILE_SIZE_IN_BYTES.id(),
+        record.get(Column.TOTAL_DATA_FILE_SIZE_IN_BYTES.id(), Long.class));
+    stats.set(
+        Column.POSITION_DELETE_RECORD_COUNT.id(),
+        record.get(Column.POSITION_DELETE_RECORD_COUNT.id(), Long.class));
+    stats.set(
+        Column.POSITION_DELETE_FILE_COUNT.id(),
+        record.get(Column.POSITION_DELETE_FILE_COUNT.id(), Integer.class));
+    stats.set(
+        Column.EQUALITY_DELETE_RECORD_COUNT.id(),
+        record.get(Column.EQUALITY_DELETE_RECORD_COUNT.id(), Long.class));
+    stats.set(
+        Column.EQUALITY_DELETE_FILE_COUNT.id(),
+        record.get(Column.EQUALITY_DELETE_FILE_COUNT.id(), Integer.class));
+    stats.set(
+        Column.TOTAL_RECORD_COUNT.id(), record.get(Column.TOTAL_RECORD_COUNT.id(), Long.class));
+    stats.set(Column.LAST_UPDATED_AT.id(), record.get(Column.LAST_UPDATED_AT.id(), Long.class));
+    stats.set(
+        Column.LAST_UPDATED_SNAPSHOT_ID.id(),
+        record.get(Column.LAST_UPDATED_SNAPSHOT_ID.id(), Long.class));
+
+    return PartitionStatsRecord.create(record.struct(), stats);
+  }
+
+  @VisibleForTesting
+  static Iterator<PartitionStatsRecord> statsToRecords(
```

Comment:
This method is not called from unit tests, @VisibleForTesting seems to be not needed.

Reply:
removed